exec/java-exec/src/main/java/org/apache/drill/exec/expr/stat/ParquetBooleanPredicate.java

```diff
@@ -46,15 +46,32 @@ private ParquetBooleanPredicate(String name, List<LogicalExpression> args, Expre
       ExpressionPosition pos
   ) {
     return new ParquetBooleanPredicate<C>(name, args, pos) {
+      /**
+       * Evaluates a compound "AND" filter on the statistics of a RowGroup (the filter reads "filterA and filterB").
+       * Return value :<ul>
+       *   <li>ALL : only if all filters return ALL
+       *   <li>NONE : if one filter at least returns NONE
+       *   <li>SOME : all other cases
+       * </ul>
+       */
       @Override
-      public boolean canDrop(RangeExprEvaluator<C> evaluator) {
-        // "and" : as long as one branch is OK to drop, we can drop it.
+      public RowsMatch matches(RangeExprEvaluator<C> evaluator) {
+        RowsMatch childMatch;
+        RowsMatch resultMatch = RowsMatch.NONE;
```

Comment:
Should `resultMatch` be initialized to `RowsMatch.ALL` so it can be returned without the extra check on line 71?

Reply:
No. 11 test fails in this case.

Reply:
Why does it fail in this case?

Reply:
Probably we don't enter the for loop... resultMatch take into account that one at least o fthe rowgroups is in SOME state, which is enough to NOT drop the filter at all.

Reply:
Finally, I can suppress a few lines:
RowsMatch childMatch;
        RowsMatch resultMatch = RowsMatch.NONE;
        for (LogicalExpression child : this) {
          if (child instanceof ParquetFilterPredicate) {
            childMatch = ((ParquetFilterPredicate) child).matches(evaluator);
            if (childMatch == RowsMatch.ALL) {
              return RowsMatch.ALL;  // One at least is ALL => can drop filter but not RG
            }
            if (childMatch == RowsMatch.SOME) {
              resultMatch = RowsMatch.SOME;
            }
          }
        }
        return resultMatch;

Reply:
And it works also with your first guess : resultMatch be initialized to RowsMatch.ALL
:)

Reply:
I'd suggest that you step through the code in a debugger for any test case that fails and check why it fails. For the case when `this` is empty and for loop is skipped, if `resultMatch` is initialized to `RowsMatch.NONE`, the condition on line 71 won't be true and you will return `RowsMatch.ALL`. The same result will be returned if `resultMatch` is initialized to `RowsMatch.ALL` and the check on line 71 is removed unless I miss something.

Reply:
Agree. Which can now be seen on each matches() function.

Reply:
Last but not least :), I can rewrite the functions using the switch syntax:
public RowsMatch matches(RangeExprEvaluator<C> evaluator) {
        RowsMatch resultMatch = RowsMatch.NONE;
        for (LogicalExpression child : this) {
          if (child instanceof ParquetFilterPredicate) {
            switch (((ParquetFilterPredicate) child).matches(evaluator)) {
              case ALL :
                return RowsMatch.ALL;  // One at least is ALL => can drop filter but not RG
              case SOME :
                resultMatch = RowsMatch.SOME;
              default:
                ;
            }
          }
        }
        return resultMatch;
      }

Better than previous? We get rid of childMatch variable... What do you prefer?